crates/libcgroups/src/v2/devices/controller.rs

```diff
@@ -73,20 +79,99 @@ impl Devices {
         // IMHO, this is too complicated, and in most cases, we just attach program once without
         // already attached programs.
 
+        // get the fd of the cgroup root
         let fd = nix::dir::Dir::open(
             cgroup_root.as_os_str(),
             OFlag::O_RDONLY | OFlag::O_DIRECTORY,
             Mode::from_bits(0o600).unwrap(),
         )?;
 
-        let old_progs = bpf::prog_query(fd.as_raw_fd())?;
-        bpf::prog_attach(prog_fd, fd.as_raw_fd())?;
+        // collect the programs attached to this cgroup
+        let old_progs = bpf_prog::query(fd.as_raw_fd())?;
+        // attach our new program
+        bpf_prog::attach(prog_fd, fd.as_raw_fd())?;
+        // detach all previous programs
         for old_prog in old_progs {
-            bpf::prog_detach2(old_prog.fd, fd.as_raw_fd())?;
+            bpf_prog::detach2(old_prog.fd, fd.as_raw_fd())?;
         }
 
         Ok(())
     }
 }
 
-// FIXME: add tests, but how to?
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test::setup;
+    use serial_test::serial;
+
+    use oci_spec::runtime::{LinuxDeviceCgroupBuilder, LinuxDeviceType};
+    use std::os::unix::io::RawFd;
+
+    use bpf::mock_prog;
+
+    #[test]
+    #[serial(bpf)] // mock contexts are shared
+    fn test_apply_devices() {
+        // arrange
+        let (tmp, _) = setup("test_apply_devices", "some.value");
+        let a_type = LinuxDeviceCgroupBuilder::default()
+            .typ(LinuxDeviceType::A)
+            .build()
+            .unwrap();
+        let file_descriptor: RawFd = 6;
+
+        // expect
+        let bump_memlock_rlimit = mock_prog::bump_memlock_rlimit_context();
+        let load = mock_prog::load_context();
+        let query = mock_prog::query_context();
+        let attach = mock_prog::attach_context();
+        let detach2 = mock_prog::detach2_context();
+        bump_memlock_rlimit.expect().once().returning(|| Ok(()));
+        load.expect()
+            .once()
+            .returning(move |_, _| Ok(file_descriptor));
+        query.expect().once().returning(|_| Ok(vec![]));
+        attach.expect().once().returning(|_, _| Ok(()));
+        detach2.expect().never();
+
+        // act
+        Devices::apply_devices(&tmp, &Some(vec![a_type])).expect("Could not apply devices");
+    }
+
+    #[test]
+    #[serial(bpf)] // mock contexts are shared
```

Comment:
[From the docs for serial](https://docs.rs/serial_test/latest/serial_test/attr.serial.html)

> Multiple tests with the serial attribute are guaranteed to be executed in serial. Ordering of the tests is not guaranteed however. If you want different subsets of tests to be serialised with each other, but not depend on other subsets, you can add an argument to serial, and all calls with identical arguments will be called in serial.